File: rob599_hw3/scripts/median_filter.py
#!/usr/bin/env python

import rospy
import math
import numpy as np
from sensor_msgs.msg import LaserScan
#######################################
# Laser Scan: 
# 	Header: Seq, Stamp, frame_id
# 	Angle_min, Angle_max, Angle_Increment, Time_Increment
#	Scan time, range_min, range_max, ranges, intensities 
#######################################

class Laser_Filter:
	def __init__(self):
		#rospy.on_shutdown(self.save_csv)
		self.laser_sub = rospy.Subscriber('/base_scan', LaserScan, self.laser_callback)
		self.scan_pub = rospy.Publisher('/laser_scan', LaserScan, queue_size= 1)


	def laser_callback(self, msg):
		filtered_values = LaserScan()
		distance = np.array(msg.ranges)
		filtered_values.header = msg.header
		filtered_values.angle_increment = msg.angle_increment
		filtered_values.time_increment = msg.time_increment
		filtered_values.scan_time = msg.scan_time
		filtered_values.range_min = msg.range_min
		filtered_values.range_max = msg.range_max
		filtered_values.intensities = msg.intensities
		angle = filtered_values.angle_increment
		min_angle = msg.angle_min
		max_angle = msg.angle_max
		
		median_filter_size = rospy.get_param('median_filter_size')
		if median_filter_size < 1:
			median_filter_size = 1
		elif median_filter_size > len(distance)/2 - 1:
			median_filter_size = int(len(distance)/2 - 1)

		filtered_values_ranges = np.zeros(len(distance))

		for i in range(len(distance) - median_filter_size - 1):
			if i < median_filter_size:
				filtered_values_ranges[i] = 0
			else:
				filtered_values_ranges[i] = np.median(distance[(i - median_filter_size):(i + median_filter_size+1)])
				if filtered_values_ranges[i] > msg.range_max or filtered_values_ranges[i] < 0:
					filtered_values_ranges[i] = 0

		filtered_values.ranges = filtered_values_ranges
		filtered_values.angle_min = min_angle
		filtered_values.angle_max = max_angle
		self.scan_pub.publish(filtered_values)


if __name__ == '__main__':
	rospy.init_node('median_filter', anonymous=True)
	laser_filter = Laser_Filter()
	try:
		rospy.spin()
	except KeyboardInterrupt:
		print("Shutting down")

File: rob599_hw3/src/median_filter.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "sensor_msgs/LaserScan.h"
#include <sstream>

class Laser_Filter 
{
  public:
    ros::NodeHandle n;
    ros::Publisher filtered_laser_pub = n.advertise<sensor_msgs::LaserScan>("/laser_scan", 1000);
    void laser_Callback(const sensor_msgs::LaserScan::ConstPtr& msg);
};


void Laser_Filter::laser_Callback(const sensor_msgs::LaserScan::ConstPtr& msg)
{
  int filtered_value_len = sizeof(msg ->ranges)/sizeof(msg ->ranges[0]);
  std::vector<float> filtered_values_ranges(filtered_value_len, 0);
  
  sensor_msgs::LaserScan filtered_scan;
  filtered_scan.header = msg->header;
  filtered_scan.angle_max = msg->angle_max;
  filtered_scan.angle_min = msg->angle_min;
  filtered_scan.ranges = msg->ranges;
  filtered_scan.angle_increment = msg->angle_increment;
  filtered_scan.time_increment = msg->time_increment;
  filtered_scan.scan_time = msg->scan_time;
  filtered_scan.intensities = msg->intensities;
  filtered_scan.range_min = msg->range_min;
  filtered_scan.range_max = msg->range_max;

  int median_filter_size;
  n.getParam("median_filter_size", median_filter_size);
  if (median_filter_size < 1)
  {
    median_filter_size = 1;
  }
  else if(median_filter_size > filtered_value_len/2 - 1)
  {
    median_filter_size = filtered_value_len/2 - 1;
  }
  int sizee = 2*median_filter_size+1;
  for(int i=0; i<filtered_value_len - median_filter_size -1; i++)
  {
    if(i<median_filter_size)
    {
      float not_a_median;
      filtered_values_ranges[i] = not_a_median;
    }
    else{
         std::vector<float> median_array(sizee, 0);
         int count = 0;
        for(int j= i - median_filter_size; j<sizee; j++)
        {
         median_array[count] = filtered_scan.ranges[j];
         count++;
        }
        std::sort(median_array.begin(), median_array.end());
        int median_idx = sizee/2;
        filtered_values_ranges[i] = median_array[median_idx];
        }
  }
  for(int i =0; i <filtered_value_len; i++)
  {
  filtered_scan.ranges[i] = filtered_values_ranges[i];
  }
  filtered_laser_pub.publish(filtered_scan);
  }


int main(int argc, char **argv)
{
  ros::init(argc, argv, "median_filter");

  ros::NodeHandle n;

  Laser_Filter laser;
  ros::Subscriber sub = n.subscribe<sensor_msgs::LaserScan>("/base_scan", 1000, &Laser_Filter::laser_Callback, &laser);

  ros::spin();

  return 0;
}